vcd/resource_vcd_vapp_vm_tools.go

```diff
@@ -0,0 +1,1168 @@
+func isPrimaryNicRemoved(d *schema.ResourceData) bool {
+	_, newNetworkRaw := d.GetChange("network")
+	newNetworks := newNetworkRaw.([]interface{})
+
+	var foundPrimaryNic bool
+	for _, newNet := range newNetworks {
+		netMap := newNet.(map[string]interface{})
+		isPrimary := netMap["is_primary"].(bool)
+		if isPrimary {
+			foundPrimaryNic = true
+			break
+		}
+	}
+
+	return !foundPrimaryNic
+}
+
+func updateVmSpecSection(vmSpecSection *types.VmSpecSection, vm *govcd.VM, description string) error {
+	// add missing values if not inherited from template, otherwise API throws error if some value is nil
+	if vmSpecSection.MemoryResourceMb.Reservation == nil {
+		vmSpecSection.MemoryResourceMb.Reservation = takeInt64Pointer(int64(0))
+	}
+	if vmSpecSection.MemoryResourceMb.Limit == nil {
+		vmSpecSection.MemoryResourceMb.Limit = takeInt64Pointer(int64(-1))
+	}
+	if vmSpecSection.MemoryResourceMb.SharesLevel == "" {
+		vmSpecSection.MemoryResourceMb.SharesLevel = "NORMAL"
+	}
+	if vmSpecSection.CpuResourceMhz.Reservation == nil {
+		vmSpecSection.CpuResourceMhz.Reservation = takeInt64Pointer(int64(0))
+	}
+	if vmSpecSection.CpuResourceMhz.Limit == nil {
+		vmSpecSection.CpuResourceMhz.Limit = takeInt64Pointer(int64(-1))
+	}
+	if vmSpecSection.CpuResourceMhz.SharesLevel == "" {
+		vmSpecSection.CpuResourceMhz.SharesLevel = "NORMAL"
+	}
+	_, err := vm.UpdateVmSpecSection(vmSpecSection, description)
+	if err != nil {
+		return fmt.Errorf("error updating Vm Spec Section: %s", err)
+	}
+	return nil
+}
+
+func getVmFromResource(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*VCDClient, *govcd.Org, *govcd.Vdc, *govcd.VApp, string, *govcd.VM, error) {
+	vcdClient := meta.(*VCDClient)
+
+	org, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return nil, nil, nil, nil, "", nil, fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vappName := d.Get("vapp_name").(string)
+	vapp, err := vdc.GetVAppByName(vappName, false)
+
+	if err != nil {
+		additionalMessage := ""
+		if vmType == standaloneVmType {
+			additionalMessage = fmt.Sprintf("\nAdding a vApp name to a standalone VM is not allowed." +
+				"Please use 'vcd_vapp_vm' resource to specify vApp")
+			dSet(d, "vapp_name", "")
+		}
+
+		return nil, nil, nil, nil, "", nil, fmt.Errorf("[getVmFromResource] error finding vApp '%s': %s%s", vappName, err, additionalMessage)
+	}
+
+	identifier := d.Id()
+	if identifier == "" {
+		identifier = d.Get("name").(string)
+	}
+	if identifier == "" {
+		return nil, nil, nil, nil, "", nil, fmt.Errorf("[VM update] neither name or ID was set")
+	}
+
+	vm, err := vapp.GetVMByNameOrId(identifier, false)
+
+	if err != nil {
+		d.SetId("")
+		return nil, nil, nil, nil, "", nil, fmt.Errorf("[VM update] error getting VM %s: %s", identifier, err)
+	}
+	return vcdClient, org, vdc, vapp, identifier, vm, nil
+}
+
+// updates attached disks to latest state. Removed not needed and add new ones
```

Comment:
```suggestion
// attachDetachIndependentDisks updates attached disks to latest state, removes not needed, and adds new ones
```
The suggestion is based on my understanding. Please rephrase the comment if I got it wrong/

Reply:
Yes. This is correct assumption. Changed.